### PR TITLE
AP_Camera: fixed CAM_MIN_INTERVAL

### PR DESCRIPTION
**Pull request by tridge**
delay the next photo until minimum interval is met, which is what the documentation says. This fixes a nasty bug with mission plans where an extra photo can be triggered by a camera trigger in a mission which results in the number of CAM msgs being more than the number of images on the microSD, which makes the mapping run unusable

### CODE REVIEW

File: libraries/AP_Camera/AP_Camera.h [hunk 1]
```diff
@@ -127,6 +127,7 @@ class AP_Camera {
     AP_Int16        _min_interval;      // Minimum time between shots required by camera
     AP_Int16        _max_roll;          // Maximum acceptable roll angle when trigging camera
     uint32_t        _last_photo_time;   // last time a photo was taken
+    bool            _trigger_pending;
```
Comment (rmackay9) [inline libraries/AP_Camera/AP_Camera.h:130]: maybe add a comment beside this variable's declaration to match the others..

**Comment by rmackay9** [comment]
This looks OK to me.

**Comment by WickedShell** [comment]
> which results in the number of CAM msgs being more than the number of images on the microSD, which makes the mapping run unusable

This is something that GCS's can handle far better. Any handling software that expects a 1:1 correlation of photos on the card to CAM messages is doomed to failure. I've pushed documentation out in the past (on diydrones actually) for how to handle this better, but the knowledge generally seems to have been missed. It does require slightly more input from the user to set a single synchronization point, but with this you can handle dropping CAM messages.

Basically I just want to note that expecting a perfect 1:1 of CAM msgs to images is incorrect, and flawed. (Although building towards that better world is totally worth it).

**Comment by tridge** [comment]
> Basically I just want to note that expecting a perfect 1:1 of CAM msgs to images is incorrect, and flawed.

I agree. Do you happen to know of some way to get millisecond timestamps on the images? (using a sony camera). If we can get ms timestamps with exif then we could match time deltas to match images.
FAT only stores time at 2s resolution, and it looks like exif timestamps on sony cameras have 1s resolution. When taking photos at around 0.9Hz that makes good matching hard

File: libraries/AP_Camera/AP_Camera.h (hunk 1 above)
Comment (tridge) [reply libraries/AP_Camera/AP_Camera.h:130]: done

**Comment by WickedShell** [comment]
> Do you happen to know of some way to get millisecond timestamps on the images? (using a sony camera).

Sadly the best I've found is 1 second resolution. If you have access to some of the more industrial API's you can get it a bit better, but you still have a host of other problems that effectively leaves you in the same boat.

The path I've found is a way to get a timestamp offset to match a photo and proceed through pairing using that, (and assuming matches in order until outside of a defined tolerance window). In order for a long term error to not propagate you do periodically need a long space with no photos then the first photo to actually happen again, which resyncs the camera. (This tends to happen orginaically on a fixed wing system due to the wind slowing legs down). In photogrammetry applications it's also possible to detect which photos have bad stamps and correct them after the fact to be correct, as long as the majority of the photos are correct and synced, but this starts to require either more user input, or a lot more custom scripting.

All of the above reasons are why it's worth ensuring that we don't knowingly trigger more photos then can be taken.

Camera feedback is *very* important though as it helps cut a lot of the above error out of the loop. If you can wire the feedback in your system is far more reliable, and easier to tag on. FWIW my system is actually prone to taking more photos then requested, but because of the feedback it's very manageable.

**Comment by tridge** [comment]
test flown on a GriffinPro on Sunday, performed correctly
